lib/resty/openssl/x509/store_ctx.lua

```diff
@@ -0,0 +1,234 @@
+  end
+
+  local purp = C.X509_PURPOSE_get0(idx)
+  local i = C.X509_PURPOSE_get_id(purp)
+
+  if C.X509_STORE_CTX_set_purpose(self.ctx, i) ~= 1 then
+    return false, format_error("x509.store_ctx:set_purpose: X509_STORE_set_purpose")
+  end
+
+  return true
+end
+
+function _M:set_flags(...)
+  local flags = 0
+  for _, f in ipairs({...}) do
+    flags = bor(flags, f)
+  end
+
+  C.X509_STORE_CTX_set_flags(self.ctx, flags)
+
+  return true
+end
+
+function _M:add_crl(crl)
+  if not crl_lib.istype(crl) then
+    return nil, "x509.store_ctx:add_crl: expect a x509.crl instance at #3"
+  end
+
+  if not self.crl_stack then
+    local raw, err = crl_stack_new()
+    if raw == nil then
+      return nil, err
+    end
+
+    self.crl_stack = setmetatable({
+      ctx = raw,
+    }, crl_stack_mt)
+  end
+
+  local dup = C.X509_CRL_dup(crl.ctx)
+  if dup == nil then
+    return nil, format_error("x509.store_ctx:add_crl: X509_CRL_dup")
+  end
+
+  local ok, err = crl_stack_add(self.crl_stack.ctx, dup)
+  if not ok then
+    C.X509_CRL_free(dup)
+    return nil, err
+  end
+
+  return true
+end
+
+function _M:verify(return_chain)
+  if self.crl_stack then
+    C.X509_STORE_CTX_set0_crls(self.ctx, self.crl_stack.ctx)
+
+    -- auto-set the flag
+    self:set_flags(flag_crl_check)
+  end
+
+  local code = C.X509_verify_cert(self.ctx)
+  if code == 1 then -- verified
+    self.verified = true
+    if not return_chain then
+      return true, nil
+    end
+
+    local chain_ctx = x509_vfy_macro.X509_STORE_CTX_get0_chain(self.ctx)
+    return chain_lib.dup(chain_ctx)
+
+  elseif code == 0 then -- unverified
+    local vfy_code = C.X509_STORE_CTX_get_error(self.ctx)
+    return nil, ffi_str(C.X509_verify_cert_error_string(vfy_code))
+  end
+
+  -- error
+  return nil, format_error("x509.store_ctx:verify: X509_verify_cert", code)
+end
+
+function _M:check_revocation(verified_chain)
+  if not OPENSSL_11_OR_LATER then
+    return nil, "x509.store_ctx:check_revocation: this API is supported from OpenSSL 1.1.0"
+  end
+
+  if verified_chain then
+    if not chain_lib.istype(verified_chain) then
+      return nil, "x509.store_ctx:check_revocation: expect a x509.chain instance at #1"
+    end
+
+    C.X509_STORE_CTX_set0_verified_chain(self.ctx, verified_chain.ctx)
+  else
+    if not self.verified then
+      return nil, "x509.store_ctx:check_revocation: store_ctx isn't verified and verified_chain argument isn't passed"
+    end
+  end
+
+  if self.crl_stack then
+    C.X509_STORE_CTX_set0_crls(self.ctx, self.crl_stack.ctx)
+  end
+
+  -- enables CRL checking for the certificate chain leaf certificate.
+  -- An error occurs if a suitable CRL cannot be found.
+  -- Note: this does not check for certificates in the chain.
+  -- We just want to check the revocation of leaf certificate here,
+  -- so there's no need to verify the complete chain.
+  self:set_flags(flag_partial_chain + flag_crl_check)
+
+  local check_revocation = ffi.cast("int (*)(X509_STORE_CTX *)",
```

Comment:
`check_revocation` needs a explict `check_revocation:free()`

Reply:
Thanks for catching this!

Reply:
Seems we don't need `ffi.cast` here at all. Misled by ChatGPT😂. It's used when casting a Lua function to a C function pointer